### DISCUSSION TITLE
Infer Literal Value Type of Const List Elements

### DISCUSSION THREAD

**Discussion by darwintree** [microsoft/pyright#5340] [Q&A] [answered]
Here is the example code:

![image](https://github.com/microsoft/pyright/assets/17946284/119d9b81-c608-429a-be0b-a97b13743952)

```python
from typing import Literal

def accept_literal(val: Literal["hello", "world"]):
    print(val)
    
for val in ["hello", "world"]:
    accept_literal(val)
```

```
Argument of type "str" cannot be assigned to parameter "val" of type "Literal['hello', 'world']" in function "accept_literal"
  Type "str" cannot be assigned to type "Literal['hello', 'world']"
    "str" cannot be assigned to type "Literal['hello']"
    "str" cannot be assigned to type "Literal['world']"
```

It is natural that pylance inferring val as `str` type, but it brings about the problem in the above code. I wonder if this the intended behaviour? And is there any elegant way to fix the above problem? Thank you

**Comment by erictraut** [comment, marked as answer]
This is intended. The type of the expression `["hello", "world"]` is `list[str]`. That means `val` receives a `str` value.

It will work if you use a tuple instead of a list. The type of expression `("hello", "world")` is `tuple[Literal["hello"], Literal["world"]]`, so `val` receives a `Literal["hello", "world"]` value. This won't work in mypy because its tuple inference doesn't retain literal values.

The other way you could do this is to define a separate variable for a list that holds only literal values:

```python
my_list: list[Literal["hello", "world"]] = ["hello", "world"]
for val in my_list:
    accept_literal(val)
```
